### PR TITLE
scalarize shape computation ops on tensor level whenever possible

### PR DESCRIPTION
**Pull request by wyzero**

### CODE REVIEW

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc [hunk 1]
```diff
@@ -301,13 +316,181 @@ struct ExtractElementOfTieShapeOpCanonicalizationPattern
...
+// to:
+//   %2 = tensor.extract %0[] : ...
+struct ExtractElementOfReshapeOpCanonicalizationPattern
+    : public OpRewritePattern<tensor::ExtractOp> {
+  using OpRewritePattern<tensor::ExtractOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(tensor::ExtractOp op,
+                                PatternRewriter& rewriter) const override {
+    auto tensorTy = op.tensor().getType().dyn_cast<RankedTensorType>();
+    if (!tensorTy || !tensorTy.hasStaticShape() ||
```
Comment (qiuxiafei) [inline tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc:407]: why it's different from `isCandidateShapeTensorType`?

Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc:407]: Reshape is more special in our use cases. For example:

```
%0 = reshape from tensor<i32> to tensor<1xi32>
%1 = reshape from tensor<i32> to tensor<1xi32>
%2 = mhlo.concat(%0, %1) : tensor<2xi32>
```
Here the input of reshape has rank zero, which is usually not considered as a shape tensor. Thus, we need to relax the restrictions of reshape here.

File: tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc [hunk 2]
```diff
@@ -301,13 +316,181 @@ struct ExtractElementOfTieShapeOpCanonicalizationPattern
...
+    auto indexOp = dyn_cast_or_null<arith::ConstantOp>(
+        op.indices().front().getDefiningOp());
+    if (!indexOp) return failure();
+    int64_t index = indexOp.getValue().cast<IntegerAttr>().getInt();
+
+    auto concatOp = op.tensor().getDefiningOp<mhlo::ConcatenateOp>();
+    if (!concatOp) return failure();
+    if (!isCandidateShapeTensorType(op.tensor().getType())) return failure();
+
+    for (auto& en : llvm::enumerate(concatOp->getOperands())) {
```
[the remark this answers is not in the document]
Comment (wyzero) [reply tao_compiler/mlir/disc/transforms/disc_shape_optimization.cc]: done

**Review by qiuxiafei — APPROVED** [review]

**Comment by wyzero** [comment]
to #325

**Pull request merged**